examples/component-program/main.go

```diff
@@ -0,0 +1,36 @@
+// Copyright 2016-2025, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// package main shows how a simple comoponent provider can be created using existing
+// Pulumi programs that contain components.
+package main
+
+import (
+	"github.com/pulumi/pulumi-go-provider/component"
+	"github.com/pulumi/pulumi-go-provider/examples/component-program/nested"
+)
+
+// We need to register all the component resource types we want to expose during initialization.
+func init() {
```

Comment:
Why does it need to happen in init? Seems like they could be run in main anytime before invoking `component.ProviderHost`?

Reply:
Yes, the registrations could also just happen in the main function, as long as it occurs before invoking `component.ProviderHost`.

I've opted to showcase this within the `init` function as we plan to have a separate tool to codegenerate these type registrations. This would be helpful if there are multiple component resources within a Pulumi Go program that needs to be exposed. When generating this code, it would be easiest if they occur within an `init` function.

This commit (https://github.com/pulumi/pulumi-go-provider/pull/308/commits/a2aab8c5d453656eae3e89d33d1f4b14b131abf6) is an example of this generator being implemented. I am punting this work to be released at a later stage so we can have a more thorough design review.